cpp/src/arrow/config.cc

```diff
@@ -73,7 +76,30 @@ RuntimeInfo GetRuntimeInfo() {
       MakeSimdLevelString([&](int64_t flags) { return cpu_info->IsSupported(flags); });
   info.detected_simd_level =
       MakeSimdLevelString([&](int64_t flags) { return cpu_info->IsDetected(flags); });
+  info.using_os_timezone_db = USE_OS_TZDB;
+#if !USE_OS_TZDB
+  info.timezone_db_path = timezone_db_path;
+#endif
   return info;
 }
 
+Status Initialize(const ArrowGlobalOptions& options) noexcept {
+  if (options.tz_db_path.has_value()) {
+#if !USE_OS_TZDB
+    try {
+      arrow_vendored::date::set_install(options.tz_db_path.value());
+      arrow_vendored::date::reload_tzdb();
+    } catch (const std::runtime_error& e) {
+      return Status::ExecutionError(e.what());
+    }
+    timezone_db_path = options.tz_db_path.value();
+#else
+    return Status::Invalid(
+        "Arrow was set to use OS timezone database at compile time, so it cannot be set "
+        "at runtime.");
```

Comment:
The error message is weird, it reads like you can set to use OS timezone database at runtime.